File: algorithm/_misc2/all-o-one-data-structure.cpp
// 432. All O`one Data Structure
// Difficulty: Hard

// Implement a data structure supporting the following operations:

// Inc(Key) - Inserts a new key with value 1. Or increments an existing key by 1. Key is guaranteed to be a non-empty string.
// Dec(Key) - If Key's value is 1, remove it from the data structure. Otherwise decrements an existing key by 1. 
//            If the key does not exist, this function does nothing. Key is guaranteed to be a non-empty string.
// GetMaxKey() - Returns one of the keys with maximal value. If no element exists, return an empty string "".
// GetMinKey() - Returns one of the keys with minimal value. If no element exists, return an empty string "".
// Challenge: Perform all these in O(1) time complexity.

// Time:  O(1), per operation
// Space: O(k)

class AllOne {
public:
  /** Initialize your data structure here. */
  AllOne() {
  }

  /** Inserts a new key <Key> with value 1. Or increments an existing key by 1. */
  void inc(string key) {
    if (!_idx.count(key)) {
      _idx[key] = _buckets.insert(_buckets.begin(), {0, {key}});
    }

    auto bucket = _idx[key], next = next(bucket);
    if (next == _buckets.end() || next->value > bucket->value + 1) {
      next = _buckets.insert(next, {bucket->value + 1, {}});
    }
    next->keys.insert(key);
    _idx[key] = next;

    bucket->keys.erase(key);
    if (bucket->keys.empty()) {
      _buckets.erase(bucket);
    }
  }

  /** Decrements an existing key by 1. If Key's value is 1, remove it from the data structure. */
  void dec(string key) {
    if (!_idx.count(key)) {
      return;
    }

    auto bucket = _idx[key], prev = prev(bucket);
    _idx.erase(key);
    if (bucket->value > 1) {
      if (bucket == _buckets.begin() || prev->value < bucket->value - 1) {
        prev = _buckets.insert(bucket, {bucket->value - 1, {}});
      }
      prev->keys.insert(key);
      _idx[key] = prev;
    }

    bucket->keys.erase(key);
    if (bucket->keys.empty()) {
      _buckets.erase(bucket);
    }
  }

  /** Returns one of the keys with maximal value. */
  string getMaxKey() {
    return _buckets.empty() ? "" : *(_buckets.rbegin()->keys.begin());
  }

  /** Returns one of the keys with Minimal value. */
  string getMinKey() {
    return _buckets.empty() ? "" : *(_buckets.begin()->keys.begin());
  }

private:
  struct Bucket {
    int value;
    unordered_set<string> keys;
  };
  list<Bucket> _buckets;
  unordered_map<string, list<Bucket>::iterator> _idx;
};

/**
 * Your AllOne object will be instantiated and called as such:
 * AllOne obj = new AllOne();
 * obj.inc(key);
 * obj.dec(key);
 * string param_3 = obj.getMaxKey();
 * string param_4 = obj.getMinKey();
 */

File: data-structure-and-algorithm/typescript/leetcode/src/all-o-one-data-structure.ts
/**
Design a data structure to store the strings' count with the ability to return the strings with minimum and maximum counts.

Implement the AllOne class:

AllOne() Initializes the object of the data structure.
inc(String key) Increments the count of the string key by 1. If key does not exist in the data structure, insert it with count 1.
dec(String key) Decrements the count of the string key by 1. If the count of key is 0 after the decrement, remove it from the data structure. It is guaranteed that key exists in the data structure before the decrement.
getMaxKey() Returns one of the keySet with the maximal count. If no element exists, return an empty string "".
getMinKey() Returns one of the keySet with the minimum count. If no element exists, return an empty string "".


Example 1:

Input
["AllOne", "inc", "inc", "getMaxKey", "getMinKey", "inc", "getMaxKey", "getMinKey"]
[[], ["hello"], ["hello"], [], [], ["leet"], [], []]
Output
[null, null, null, "hello", "hello", null, "hello", "leet"]

Explanation
AllOne allOne = new AllOne();
allOne.inc("hello");
allOne.inc("hello");
allOne.getMaxKey(); // return "hello"
allOne.getMinKey(); // return "hello"
allOne.inc("leet");
allOne.getMaxKey(); // return "hello"
allOne.getMinKey(); // return "leet"


Constraints:

1 <= key.length <= 10
key consists of lowercase English letters.
It is guaranteed that for each call to dec, key is existing in the data structure.
At most 5 * 104 calls will be made to inc, dec, getMaxKey, and getMinKey.
*/

import {Data} from './util/object';
import {
  DoublyLinkedList,
  DoublyLinkedListNode,
} from './data-structure/DoublyLinkedList';
import {timingSafeEqual} from 'crypto';

class AllOne {
  bucketList: DoublyLinkedList;
  keyBucketMap: Map<string, DoublyLinkedListNode>;
  constructor() {
    this.bucketList = new DoublyLinkedList();
    this.keyBucketMap = new Map<string, DoublyLinkedListNode>();
  }

  inc(key: string): void {
    console.log('\ninc');
    if (!this.keyBucketMap.has(key)) {
      const bucket = new Bucket(0, key);
      const node = this.bucketList.prepend(bucket)!;
      this.keyBucketMap.set(key, node);
    }

    const curr = this.keyBucketMap.get(key)!;
    let next = curr.next;
    if (next === null || next.data.data > curr.data.data + 1) {
      const bucket = new Bucket(curr.data.data + 1);
      next = this.bucketList.insert(curr, bucket)!;
    }

    next.data.keySet.add(key);
    this.keyBucketMap.set(key, next);

    console.log('insert');
    this.bucketList.printLog();

    this.remove(key, curr);

    console.log('remove');
    this.bucketList.printLog();
  }

  dec(key: string): void {
    console.log('\ndec');
    if (!this.keyBucketMap.has(key)) {
      return;
    }

    const curr = this.keyBucketMap.get(key)!;
    let prev = curr.prev;
    if (curr.data.data > 1) {
      if (prev === null || prev.data.data < curr.data.data - 1) {
        const bucket = new Bucket(curr.data.data - 1);
        prev = this.bucketList.insert(prev, bucket)!;
      }

      prev.data.keySet.add(key);
      this.keyBucketMap.set(key, prev);

      console.log('insert');
      this.bucketList.printLog();
    } else {
      this.keyBucketMap.delete(key);
    }

    this.remove(key, curr);

    console.log('remove');
    this.bucketList.printLog();
  }

  getMaxKey() {
    return this.bucketList.length
      ? this.bucketList.tail!.data.keySet[Symbol.iterator]().next().value
      : '';
  }

  getMinKey() {
    return this.bucketList.length
      ? this.bucketList.head!.data.keySet[Symbol.iterator]().next().value
      : '';
  }

  remove(key: string, node: DoublyLinkedListNode) {
    node.data.keySet.delete(key);
    if (node.data.keySet.size === 0) {
      this.bucketList.remove(node);
    }
  }
}

class Bucket {
  data: Data;
  keySet: Set<string>;
  constructor(data: Data, ...keyList: string[]) {
    this.data = data;
    this.keySet = new Set([...keyList]);
  }
}

const allOne = new AllOne();
allOne.inc('a');
allOne.inc('a');
allOne.inc('a');
allOne.inc('b');
allOne.dec('b');
allOne.inc('b');
allOne.dec('b');
console.log(allOne.getMaxKey());
console.log(allOne.getMinKey());
